spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/sql/CloudSqlEnvironmentPostProcessor.java

```diff
@@ -0,0 +1,203 @@
+			return false;
+		}
+	}
+
+	private CloudSqlJdbcInfoProvider buildCloudSqlJdbcInfoProvider(ConfigurableEnvironment environment, DatabaseType databaseType) {
+		CloudSqlJdbcInfoProvider cloudSqlJdbcInfoProvider = new DefaultCloudSqlJdbcInfoProvider(
+				getSqlProperty(environment, "database-name", null),
+				getSqlProperty(environment, "instance-connection-name", null),
+				getSqlProperty(environment, "ip-types", null),
+				databaseType);
+		if (LOGGER.isInfoEnabled()) {
+			LOGGER.info("Default " + databaseType.name()
+					+ " JdbcUrl provider. Connecting to "
+					+ cloudSqlJdbcInfoProvider.getJdbcUrl() + " with driver "
+					+ cloudSqlJdbcInfoProvider.getJdbcDriverClass());
+		}
+		return cloudSqlJdbcInfoProvider;
+	}
+
+	private String getSqlProperty(ConfigurableEnvironment environment, String shortName, String defaultValue) {
+		return environment.getProperty(CLOUD_SQL_PROPERTIES_PREFIX + shortName, defaultValue);
+	}
+
+	private void setCredentials(ConfigurableEnvironment environment, SpringApplication application) {
+		String encodedKey = getSqlProperty(environment, "credentials.encoded-key", null);
+		if (encodedKey != null) {
+			setCredentialsEncodedKeyProperty(encodedKey);
+		}
+		else {
+			setCredentialsFileProperty(environment, application);
+		}
+	}
+
+	private void setCredentialsEncodedKeyProperty(String encodedKey) {
+		System.setProperty(SqlCredentialFactory.CREDENTIAL_ENCODED_KEY_PROPERTY_NAME,
+				encodedKey);
+
+		System.setProperty(CredentialFactory.CREDENTIAL_FACTORY_PROPERTY,
+				SqlCredentialFactory.class.getName());
+	}
+
+	/**
+	 * Set credentials to be used by the Google Cloud SQL socket factory.
+	 *
+	 * <p>The only way to pass a {@link CredentialFactory} to the socket factory is by passing a
+	 * class name through a system property. The socket factory creates an instance of
+	 * {@link CredentialFactory} using reflection without any arguments. Because of that, the
+	 * credential location needs to be stored somewhere where the class can read it without
+	 * any context. It could be possible to pass in a Spring context to
+	 * {@link SqlCredentialFactory}, but this is a tricky solution that needs some thinking
+	 * about.
+	 *
+	 * <p>If user didn't specify credentials, the socket factory already does the right thing by
+	 * using the application default credentials by default. So we don't need to do anything.
+	 */
+	private void setCredentialsFileProperty(ConfigurableEnvironment environment, SpringApplication application) {
+		File credentialsLocationFile;
+
+		try {
+			String sqlCredentialsLocation = getSqlProperty(environment, "credentials.location", null);
+			String globalCredentialsLocation = environment.getProperty("spring.cloud.gcp.credentials.location", (String) null);
+			// First tries the SQL configuration credential.
+			if (sqlCredentialsLocation != null) {
+				credentialsLocationFile = application.getResourceLoader().getResource(sqlCredentialsLocation).getFile();
```

Comment:
Can `getResource()` return null?

Reply:
Javadoc says it's never `null`, but `getFile()` can throw an `IOException` which I guess is what we want anyway.

Reply:
You can use `Resource#exists` or `Resource#isFile`.

Reply:
I think we're fine with `IOException` here.